src/cascadia/Remoting/Monarch.cpp

```diff
@@ -130,63 +139,239 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
     }
 
-    uint64_t Monarch::_getMostRecentPeasantID()
+    // Method Description:
+    // - Retrieves the ID of the MRU peasant window. If requested, will limit
+    //   the search to windows that are on the current desktop.
+    // Arguments:
+    // - limitToCurrentDesktop: if true, only return the MRU peasant that's
+    //   actually on the current desktop.
+    // Return Value:
+    // - the ID of the most recent peasant, otherwise 0 if we could not find one.
+    uint64_t Monarch::_getMostRecentPeasantID(const bool limitToCurrentDesktop)
     {
-        if (_mostRecentPeasant != 0)
+        if (_mruPeasants.empty())
         {
-            return _mostRecentPeasant;
+            // We haven't yet been told the MRU peasant. Just use the first one.
+            // This is just gonna be a random one, but really shouldn't happen
+            // in practice. The WindowManager should set the MRU peasant
+            // immediately as soon as it creates the monarch/peasant for the
+            // first window.
+            if (_peasants.size() > 0)
+            {
+                try
+                {
+                    return _peasants.begin()->second.GetID();
+                }
+                catch (...)
+                {
+                    // This shouldn't really happen. If we're the monarch, then the
+                    // first peasant should also _be us_. So we should be able to
+                    // get our own ID.
+                    return 0;
+                }
+            }
+
+            TraceLoggingWrite(g_hRemotingProvider,
+                              "Monarch_getMostRecentPeasantID_NoPeasants",
+                              TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE));
+            return 0;
         }
 
-        // We haven't yet been told the MRU peasant. Just use the first one.
-        // This is just gonna be a random one, but really shouldn't happen
-        // in practice. The WindowManager should set the MRU peasant
-        // immediately as soon as it creates the monarch/peasant for the
-        // first window.
-        if (_peasants.size() > 0)
+        // Here, there's at least one MRU peasant.
+        //
+        // We're going to iterate over these peasants toll we find one that both:
+        // 1. Is alive
+        // 2. Meets our selection criteria (do we care if it is on this desktop?)
+        //
+        // If the peasant is dead, then we'll remove it, and try the next one.
+        // Once we find one that's alive, we'll either:
+        // * check if we only want a peasant on the current desktop, and if so,
+        //   check if this peasant is on the current desktop.
+        //   - If it isn't on the current desktop, we'll loop again, on the
+        //     following peasant.
+        // * If we don't care, then we'll just return that one.
+
+        int positionInList = 0;
+        while (_mruPeasants.begin() + positionInList < _mruPeasants.end())

```

Comment:
nit: a little note in the comment as to why an iterator is not suitable would go a long way :smile:

Reply:
nit: can these be `cbegin` and `cend`? we're not mutating through the iterators

Reply:
(and below)